src/main/java/com/google/sps/utility/GmailUtility.java

```diff
@@ -0,0 +1,158 @@
+   *     or null if either one of the arguments are invalid Trailing space added to properties so
+   *     multiple queries can be concatenated
+   */
+  public static String emailAgeQuery(int emailAge, String emailAgeUnits) {
+    // newer_than:#d where # is an integer will specify to only return emails from last # days
+    if (emailAge > 0 && (emailAgeUnits.equals("h") || emailAgeUnits.equals("d"))) {
+      return String.format("newer_than:%d%s ", emailAge, emailAgeUnits);
+    } else if (emailAge == 0 && emailAgeUnits.isEmpty()) {
+      return "";
+    }
+
+    // Input invalid
+    return null;
+  }
+
+  /**
+   * Creates Gmail query for unread emails
+   *
+   * @param unreadOnly true if only unread emails, false otherwise
+   * @return string to use in gmail (either client or API) to find emails that match these criteria
+   *     Trailing space added to properties so multiple queries can be concatenated
+   */
+  public static String unreadEmailQuery(boolean unreadOnly) {
+    // is:unread will return only unread emails
+    return unreadOnly ? "is:unread " : "";
+  }
+
+  /**
+   * Creates Gmail query to find emails from specific sender
+   *
+   * @param from email address of the sender. "" if not specified
+   * @return string to use in Gmail (either client or API) to find emails that match these criteria
+   *     Trailing space added to properties so multiple queries can be concatenated
+   */
+  public static String fromEmailQuery(String from) {
+    // from: <emailAddress> will return only emails from that sender
+    return !from.equals("") ? String.format("from:%s ", from) : "";
+  }
+
+  /**
+   * Get a gmail service instance given a credential. Required to access any Gmail API services
+   *
+   * @param credential Google Credential object with user's accessToken inside
+   * @return Gmail service instance
+   */
+  public static Gmail getGmailService(Credential credential) {
+    JsonFactory jsonFactory = AuthenticationUtility.JSON_FACTORY;
+    HttpTransport transport = AuthenticationUtility.HTTP_TRANSPORT;
+    String applicationName = AuthenticationUtility.APPLICATION_NAME;
+
+    return new Gmail.Builder(transport, jsonFactory, credential)
+        .setApplicationName(applicationName)
+        .build();
+  }
+
+  /**
+   * Lists all emails (that match the criteria, if provided) in the user's Gmail account
+   *
+   * @param gmailService instance of Gmail Service with valid credential
+   * @param query conditions applied to the search to limit which emails are returned. "" for no
+   *     restrictions.
+   * @return List of Message objects that contain a message and thread ID
+   * @throws IOException if an issue occurs with the Gmail service
+   */
+  public static List<Message> listUserMessages(Gmail gmailService, String query)
+      throws IOException {
+    List<Message> messages =
+        gmailService.users().messages().list("me").setQ(query).execute().getMessages();
```

Comment:
I wonder if the first item of this list is indeed linked to an email. I have no emails in my test account, but I'm still presented with one ID when I access https://21-dot-step7-2020.uc.r.appspot.com/

Reply:
Hmm - can you take a look at your test account's gmail account? I haven't added any to mine but I do have an email with the subject line "Finish setting up your new Google Account" and 3 others, leaving 4 entries on the panel.

Reply:
Fixed this up! Should handle the null case fine as well.